src/ray/core_worker/core_worker.cc

```diff
@@ -271,16 +274,19 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
             new rpc::CoreWorkerClient(addr, *client_call_manager_));
       });
 
-  // Register a callback to monitor removed nodes.
+  // Register a callback to monitor add/removed nodes.
   // Note we capture a shared ownership of reference_counter_
   // here to avoid destruction order fiasco between gcs_client and reference_counter_.
-  auto on_node_change = [reference_counter = this->reference_counter_](
+  auto on_node_change = [this, reference_counter = this->reference_counter_](
                             const NodeID &node_id, const rpc::GcsNodeInfo &data) {
     if (data.state() == rpc::GcsNodeInfo::DEAD) {
       RAY_LOG(INFO) << "Node failure from " << node_id
                     << ". All objects pinned on that node will be lost if object "
                        "reconstruction is not enabled.";
       reference_counter->ResetObjectsOnRemovedNode(node_id);
+      num_alive_nodes_--;
+    } else {
+      num_alive_nodes_++;
```

Comment:
How confident are you that this is a safe assumption? Might be a good idea to add a debug or info log when we increase or decrease num_alive_nodes; that way it'll be easier to see strange behavior in our scheduler if the invariant `(data.state() == rpc::GcsNodeInfo::DEAD) <-> num alive nodes goes down by one` changes.

Reply:
I think there's benefit to wrapping this in a function, or a separate class. That way we can assert that this is never below zero without bloating this callback. Anything to assert that num_alive_nodes is within reasonable bounds will help with debugging if there is weird scheduling behavior.